internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_environment", func() *schema.Resource {
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage GitLab environments.",
+
+		CreateContext: resourceGitlabProjectEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectEnvironmentRead,
+		UpdateContext: resourceGitlabProjectEnvironmentUpdate,
+		DeleteContext: resourceGitlabProjectEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "The ID or full path of the project which the environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"name": {
+				Description:  "The name of the environment",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"external_url": {
+				Description:  "Place to link to for this environment",
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"state": {
+				Description: "State the environment is in. Accepted values: available or stopped.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	name := d.Get("name").(string)
+	options := gitlab.CreateEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
+	}
+
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab environment %q", project, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.CreateEnvironment(project, &options)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			return diag.Errorf("feature Environments is not available")
+		}
+		return diag.FromErr(err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	if err == nil {
+		fmt.Printf("%T \n %v", environmentIDInt, environmentIDInt)
+	}
```

Comment:
What's going on here?

```suggestion
	if err != nil {
		return diag.Errorf("failed to parse environment id %s to int: %v", environmentID, err)
	}
```

Reply:
I think it was bad logic from the previous MR implementer.  I have changed it to what you have proposed